### PR TITLE
Expose stream application error in FFI

### LINKED ISSUE #1699: Stream application error codes are not exposed to FFI
Using quiche via FFI will sometimes return `Error::StreamReset` or `Error::StreamStopped`. Which is just `-15` or `-16`, without the attached application error code; these errors are generally returned with `quiche_conn_stream_send()` and `quiche_conn_stream_recv()`.

I'd expect these functions to optionally be able to return the application error code on connection end.

I would do a PR, but I don't know about the ABI compatibility guarantees, my favourite way of implementation would be just adding an additional `uint64_t *error_code` parameter, which is assigned to if the error is `StreamReset` or `StreamStopped`.
If ABI compatibility is important, we'll need a new function. I honestly don't have a great idea of a name for these though. `quiche_conn_stream_send_ex()`? `quiche_conn_stream_send2()`? `quiche_conn_stream_send_with_optional_error()`? I don't know.

### PR DESCRIPTION
**Pull request by bwoebi**
Fixes #1699.

As discussed in that issue, it's fine to change the API.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@ghedo any chance to get this one in ?

File: quiche/include/quiche.h [hunk 1]
```diff
@@ -373,11 +373,13 @@ size_t quiche_conn_send_quantum_on_path(const quiche_conn *conn,
...
 ssize_t quiche_conn_stream_recv(quiche_conn *conn, uint64_t stream_id,
-                                uint8_t *out, size_t buf_len, bool *fin);
+                                uint8_t *out, size_t buf_len, bool *fin,
+                                uint64_t *application_error);
 
 // Writes data to a stream.
 ssize_t quiche_conn_stream_send(quiche_conn *conn, uint64_t stream_id,
-                                const uint8_t *buf, size_t buf_len, bool fin);
+                                const uint8_t *buf, size_t buf_len, bool fin,
+                                uint64_t application_error);
```
Comment (LPardue) [inline quiche/include/quiche.h:382]:
```suggestion
                                uint64_t *out_error_code);
```

File: quiche/src/ffi.rs [hunk 2]
```diff
@@ -812,7 +812,7 @@ pub extern fn quiche_conn_send_on_path(
 #[no_mangle]
 pub extern fn quiche_conn_stream_recv(
     conn: &mut Connection, stream_id: u64, out: *mut u8, out_len: size_t,
-    fin: &mut bool,
+    fin: &mut bool, application_error: *mut u64,
```
Comment (LPardue) [inline quiche/src/ffi.rs:815]:
```suggestion
    fin: &mut bool, out_error_code: &mut u64,
```

File: quiche/src/ffi.rs [hunk 3]
```diff
@@ -834,7 +841,7 @@ pub extern fn quiche_conn_stream_recv(
 #[no_mangle]
 pub extern fn quiche_conn_stream_send(
     conn: &mut Connection, stream_id: u64, buf: *const u8, buf_len: size_t,
-    fin: bool,
+    fin: bool, application_error: *mut u64,
```
Comment (LPardue) [inline quiche/src/ffi.rs:844]:
```suggestion
    fin: bool, out_error_code: &mut u64,
```

File: quiche/src/ffi.rs [hunk 4]
```diff
@@ -823,7 +823,14 @@ pub extern fn quiche_conn_stream_recv(
     let (out_len, out_fin) = match conn.stream_recv(stream_id, out) {
         Ok(v) => v,
 
-        Err(e) => return e.to_c(),
+        Err(e) => {
+            match e {
+                Error::StreamReset(error) => unsafe { *application_error = error },
```
Comment (LPardue) [inline quiche/src/ffi.rs:828]:
```suggestion
                Error::StreamReset(error) => *out_error_code = error,
```

File: quiche/src/ffi.rs [hunk 5]
```diff
@@ -823,7 +823,14 @@ pub extern fn quiche_conn_stream_recv(
...
+                Error::StreamStopped(error) => unsafe { *application_error = error },
```
Comment (LPardue) [inline quiche/src/ffi.rs:829]:
```suggestion
                Error::StreamStopped(error) => *out_error_code = error,
```

File: quiche/src/ffi.rs [hunk 6]
```diff
@@ -845,7 +852,14 @@ pub extern fn quiche_conn_stream_send(
     match conn.stream_send(stream_id, buf, fin) {
         Ok(v) => v as ssize_t,
 
-        Err(e) => e.to_c(),
+        Err(e) => {
+            match e {
+                Error::StreamReset(error) => unsafe { *application_error = error },
```
Comment (LPardue) [inline quiche/src/ffi.rs:857]:
```suggestion
                Error::StreamReset(error) => *out_error_code = error,
```

File: quiche/src/ffi.rs [hunk 7]
```diff
@@ -845,7 +852,14 @@ pub extern fn quiche_conn_stream_send(
...
+                Error::StreamStopped(error) => unsafe { *application_error = error },
```
Comment (LPardue) [inline quiche/src/ffi.rs:858]:
```suggestion
                Error::StreamStopped(error) => *out_error_code = error,
```

File: quiche/include/quiche.h [hunk 8]
```diff
@@ -373,11 +373,13 @@ size_t quiche_conn_send_quantum_on_path(const quiche_conn *conn,
 
 // Reads contiguous data from a stream.
 ssize_t quiche_conn_stream_recv(quiche_conn *conn, uint64_t stream_id,
-                                uint8_t *out, size_t buf_len, bool *fin);
+                                uint8_t *out, size_t buf_len, bool *fin,
+                                uint64_t *application_error);
```
Comment (LPardue) [inline quiche/include/quiche.h:377]:
```suggestion
                                uint64_t *out_error_code);
```

File: quiche/include/quiche.h [hunk 9]
```diff
@@ -373,11 +373,13 @@ size_t quiche_conn_send_quantum_on_path(const quiche_conn *conn,
 
 // Reads contiguous data from a stream.
```
Comment (LPardue) [inline quiche/include/quiche.h:374]: Please update the comment to explain how out_error_code works.

File: quiche/include/quiche.h [hunk 10]
```diff
@@ -373,11 +373,13 @@ size_t quiche_conn_send_quantum_on_path(const quiche_conn *conn,
...
 ssize_t quiche_conn_stream_recv(quiche_conn *conn, uint64_t stream_id,
-                                uint8_t *out, size_t buf_len, bool *fin);
+                                uint8_t *out, size_t buf_len, bool *fin,
+                                uint64_t *application_error);
 
 // Writes data to a stream.
```
Comment (LPardue) [inline quiche/include/quiche.h:379]: Please update the comment to explain how out_error_code works.

**Review by LPardue — CHANGES_REQUESTED** [review]

File: quiche/src/ffi.rs (hunk 6 above)
Comment (bwoebi) [reply quiche/src/ffi.rs:857]: Renaming it, but cannot drop unsafe - it's dereferencing a raw pointer.

**Comment by bwoebi** [comment]
@LPardue Done :-)

File: quiche/src/ffi.rs (hunk 6 above)
Comment (ghedo) [reply quiche/src/ffi.rs]: ICYMI, the suggestion is also to change `*mut` to `&mut` which should allow removing the `unsafe`.

Comment (bwoebi) [reply quiche/src/ffi.rs]: Oh, you're right. I missed that. Changing...

File: quiche/src/ffi.rs [hunk 11]
```diff
@@ -850,7 +857,14 @@ pub extern fn quiche_conn_stream_send(
...
         Ok(v) => v as ssize_t,
 
-        Err(e) => e.to_c(),
+        Err(e) => {
+            match e {
+                Error::StreamReset(error) => *out_error_code = error,
+                Error::StreamStopped(error) => *out_error_code = error,
+                _ => {},
+            }
+            return e.to_c();
```
Comment (LPardue) [inline quiche/src/ffi.rs:866]:
```suggestion
            e.to_c()
```

**Comment by LPardue** [comment]
@bwoebi I did some rebasing and minor tweaks to get over Rust lints, however now we are seeing the C examples failing to build, which is expected due to the API change e.g.

```
client.c: In function ‘recv_cb’:
client.c:163:13: error: too few arguments to function ‘quiche_conn_stream_send’
  163 |         if (quiche_conn_stream_send(conn_io->conn, 4, r, sizeof(r), true) < 0) {
      |             ^~~~~~~~~~~~~~~~~~~~~~~
In file included from client.c:43:
../include/quiche.h:390:9: note: declared here
  390 | ssize_t quiche_conn_stream_send(quiche_conn *conn, uint64_t stream_id,
      |         ^~~~~~~~~~~~~~~~~~~~~~~
client.c:182:32: error: too few arguments to function ‘quiche_conn_stream_recv’
  182 |             ssize_t recv_len = quiche_conn_stream_recv(conn_io->conn, s,
      |                                ^~~~~~~~~~~~~~~~~~~~~~~
In file included from client.c:43:
../include/quiche.h:383:9: note: declared here
  383 | ssize_t quiche_conn_stream_recv(quiche_conn *conn, uint64_t stream_id,
      |         ^~~~~~~~~~~~~~~~~~~~~~~
```

Could you fix those up please? Then I think we'll be good to go.

**Comment by normanmaurer** [comment]
@bwoebi ping...

File: quiche/examples/server.c [hunk 12]
```diff
@@ -378,17 +378,19 @@ static void recv_cb(EV_P_ ev_io *w, int revents) {
                 fprintf(stderr, "stream %" PRIu64 " is readable\n", s);
 
                 bool fin = false;
+                uint64_t out_error_code;
```
Comment (LPardue) [inline quiche/examples/server.c:381]:
```suggestion
                uint64_t error_code;
```

**Review by LPardue — DISMISSED** [review]
LGTM now that CI is passing

File: quiche/examples/client.c [hunk 13]
```diff
@@ -160,8 +160,9 @@ static void recv_cb(EV_P_ ev_io *w, int revents) {
                 (int) app_proto_len, app_proto);
 
         const static uint8_t r[] = "GET /index.html\r\n";
-        if (quiche_conn_stream_send(conn_io->conn, 4, r, sizeof(r), true) < 0) {
-            fprintf(stderr, "failed to send HTTP request\n");
+        uint64_t error_code;
+        if (quiche_conn_stream_send(conn_io->conn, 4, r, sizeof(r), true, &error_code) < 0) {
+            fprintf(stderr, "failed to send HTTP request");
```
Comment (ghedo) [inline quiche/examples/client.c:165]:
```suggestion
            fprintf(stderr, "failed to send HTTP request: %" PRIu64 "\n", error_code);
```

**Comment by ghedo** [comment]
Merged, thanks @bwoebi!

**Pull request merged**